### PR TITLE
Enable Bitflyer potential payments

### PR DESCRIPTION
**Pull request by tsmartt**
We can now generate potential payments for Bitflyer creators. Note that I tried to clean up a little of the calling code,
as it appears we no longer need should_send_notifications. Also the class is now easily testable without using web mocks
since we can pass in the refresher.

### CODE REVIEW

File: app/services/payout/bitflyer_service.rb [hunk 1]
```diff
@@ -1,58 +1,69 @@
...
       potential_payments = []
 
-      connection = @publisher.bitflyer_connection
+      connection = publisher.bitflyer_connection
       # Sync the connection
-      # connection.sync_connection!
+      @refresher.call(bitflyer_connection: connection)
 
       potential_payments << PotentialPayment.new(
-        payout_report_id: @payout_report&.id,
```
Comment (tsmartt) [inline app/services/payout/bitflyer_service.rb:15]: @nvonpentz This is the section you are most likely interested in as it details what you'll end up seeing.

File: app/jobs/include_publisher_in_payout_report_job.rb [hunk 2]
```diff
@@ -35,15 +39,11 @@ def perform(arguments = {})
...
     when MANUAL
       potential_payment_job = Payout::ManualPayoutReportPublisherIncluder
     end
 
-    potential_payment_job.new(
-      publisher: publisher,
-      payout_report: payout_report,
-      should_send_notifications: should_send_notifications,
-    ).perform
+    potential_payment_job.new(attrs_payment_job).perform
```
Comment (yachtcaptain23) [inline app/jobs/include_publisher_in_payout_report_job.rb:47]: Would you restore using keyword arguments? That way the caller knows what's going on

Comment (tsmartt) [reply app/jobs/include_publisher_in_payout_report_job.rb:47]: Sure, I'll put it back. I used a var since I need to use it twice, but if you prefer the readability of the other, I'm not married to it.

**Comment by tsmartt** [comment]
@yachtcaptain23 we should merge https://github.com/brave-intl/publishers/pull/3206 first and then I'll rebase this one.

File: test/services/payout/bitflyer_service_test.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      subject
+    end
+
+    it 'creates potential payments for the right creator' do
+      refute_equal PotentialPayment.count, 0
+      PotentialPayment.all.each { |pp| assert_equal publisher.id, pp.publisher_id }
+    end
+
+    it 'the address is not empty' do
+      assert PotentialPayment.all.all? { |pp| pp.address.present? }
```
Comment (yachtcaptain23) [inline test/services/payout/bitflyer_service_test.rb:56]:
```suggestion
      PotentialPayment.all.each { |potential_payment| assert_equal publisher.id, potential_payment.publisher_id }
    end

    it 'the address is not empty' do
      assert PotentialPayment.all.all? { |potential_payment| potential_payment.address.present? }
```

Comment (yachtcaptain23) [reply test/services/payout/bitflyer_service_test.rb:56]: Would you be able to check to make sure the address matches rather than just if it's present?

File: app/jobs/payout/paypal_job.rb [hunk 4]
```diff
@@ -17,7 +17,6 @@ def perform(should_send_notifications: false, payout_report_id: nil, publisher_i
       IncludePublisherInPayoutReportJob.perform_async(
         payout_report_id: payout_report_id,
         publisher_id: publisher.id,
-        should_send_notifications: should_send_notifications,
```
Comment (yachtcaptain23) [inline app/jobs/payout/paypal_job.rb:20]: If you're going to remove, would you remove it everywhere in the codebase? The particular call-chain can be found with `should_only_notify`.

File: app/services/payout/bitflyer_service.rb [hunk 5]
```diff
@@ -1,58 +1,69 @@
...
+      @refresher = refresher
+      @payout_utils_class = payout_utils_class
+    end
+
+    # Change to call as soon as we refactor the other services
+    def perform(publisher:, payout_report:)
+      payout_utils = @payout_utils_class.new(payout_report: payout_report,
+                                             publisher: publisher,)
+
+      return if payout_utils.skip_publisher?
```
Comment (yachtcaptain23) [inline app/services/payout/bitflyer_service.rb:17]: I believe you need to update `Payout::Service`

File: app/jobs/include_publisher_in_payout_report_job.rb [hunk 6]
```diff
@@ -35,15 +35,14 @@ def perform(arguments = {})
     when UPHOLD
       potential_payment_job = Payout::UpholdService
     when BITFLYER
-      potential_payment_job = Payout::BitflyerService
+      return Payout::BitflyerService.build.perform(publisher: publisher,
```
Comment (yachtcaptain23) [inline app/jobs/include_publisher_in_payout_report_job.rb:38]: IMO I'd rather have all code follow one convention rather than going half-way and implementing a new pattern and any maintainer has to maintain more mental models than necessary. Furthermore, I don't see the benefit of the builder pattern when we want all code to use an inheritance pattern.

Comment (yachtcaptain23) [reply app/jobs/include_publisher_in_payout_report_job.rb:38]: Discussed offline, we'll refactor this later.

File: test/services/payout/bitflyer_service_test.rb (hunk 3 above)
Comment (tsmartt) [reply test/services/payout/bitflyer_service_test.rb:56]: Sure.

File: app/jobs/payout/paypal_job.rb (hunk 4 above)
Comment (tsmartt) [reply app/jobs/payout/paypal_job.rb:20]: I put it back, we should handle all this in a separate PR.

File: test/services/payout/bitflyer_service_test.rb (hunk 3 above)
Comment (tsmartt) [reply test/services/payout/bitflyer_service_test.rb]: Done.

File: app/services/payout/bitflyer_service.rb (hunk 5 above)
Comment (tsmartt) [reply app/services/payout/bitflyer_service.rb:17]: Done. A bunch of code was removed due to being stale.

File: app/services/payout/bitflyer_service.rb [hunk 7]
```diff
@@ -1,58 +1,69 @@
...
+
+    def initialize(refresher:, payout_utils_class:)
+      @refresher = refresher
+      @payout_utils_class = payout_utils_class
+    end
+
+    # Change to call as soon as we refactor the other services
+    def perform(publisher:, payout_report:)
+      payout_utils = @payout_utils_class.new(payout_report: payout_report,
+                                             publisher: publisher,)
```
Comment (yachtcaptain23) [inline app/services/payout/bitflyer_service.rb:15]:
```suggestion
      @payout_utils_class = payout_utils_class.new(payout_report: payout_report,
                                             publisher: publisher,)
    end

    # Change to call as soon as we refactor the other services
    def perform(publisher:, payout_report:)
```

File: app/services/payout/bitflyer_service.rb [hunk 8]
```diff
@@ -1,58 +1,69 @@
...
-          status: @publisher.last_status_update&.status,
+          suspended: publisher.suspended?,
+          status: publisher.last_status_update&.status,
           channel_stats: channel.details.stats,
           channel_type: channel.details_type
         )
       end
 
-      unless should_only_notify?
+      unless payout_utils.should_only_notify?
```
Comment (yachtcaptain23) [inline app/services/payout/bitflyer_service.rb:59]: If you take the change above, then this part will need to be refactored too.

File: app/services/payout/service.rb [hunk 9]
```diff
@@ -51,31 +51,6 @@ def skip_publisher?
...
-          create_message("Publisher is located in Japan and is not paid out through this job")
-          return true
-        end
-
-        if @publisher.paypal_connection.present?
-          create_message("Publisher has a Paypal Connection  and is not paid out through this job")
-          return true
-        end
-      end
-
```
Comment (yachtcaptain23) [inline app/services/payout/service.rb:78]: Why were these deleted? I believe we need to retain this code but also include BitFlyer to include Japanese payments.

File: test/services/payout/uphold_service_test.rb [hunk 10]
```diff
@@ -247,31 +247,13 @@ class UpholdServiceTest < ActiveJob::TestCase
     end
   end
 
-  describe "for a paypal user" do
```
Comment (yachtcaptain23) [inline test/services/payout/uphold_service_test.rb:250]: Why is this test being deleted?

File: app/services/payout/bitflyer_service.rb (hunk 7 above)
Comment (tsmartt) [reply app/services/payout/bitflyer_service.rb:15]: This would require passing in the `publisher` and `payout_report` to `build`, which should only be used for non-dynamic utility/third party modules. As the surrounding code is modularized, the utils class will no longer need to be instantiated and those methods can become `self`, effectively turning it into a module of functions.

File: app/services/payout/service.rb (hunk 9 above)
Comment (tsmartt) [reply app/services/payout/service.rb:78]: I believe these are irrelevant for 2 reasons: 
1) Paypal is going away, so we can clean up this code but more importantly 
2) As the comment suggested, this can be removed once that issue is resolved, which it was. We pay out based on the `selected_wallet_provider`, so these checks are unnecessary.

Comment (tsmartt) [reply app/services/payout/service.rb:78]: I'm going to separate out the data cleanup into a separate PR, so I'll restore these for now.

**Review by yachtcaptain23 — APPROVED** [review]
LG2M. Please remove running the paypal payouts in either this PR or in a separate PR

**Comment by tsmartt** [comment]
>  Please remove running the paypal payouts in either this PR or in a separate PR

Will do. Was planning a separate PR.
